sdk/samples/iot/paho_iot_pnp_with_provisioning_sample.c

```diff
@@ -0,0 +1,1218 @@
+// the functionality would need to match the DTMI and you would need to update the below 'model_id'.
+// Please see the sample README for more information on this DTMI.
+static az_span const model_id = AZ_SPAN_LITERAL_FROM_STR("dtmi:com:example:Thermostat;1");
+
+// Plug and Play Connection Values
+static uint32_t connection_request_id_int = 0;
+static char connection_request_id_buffer[16];
+
+// Plug and Play Property Values
+static az_span const property_success_name = AZ_SPAN_LITERAL_FROM_STR("success");
+static az_span const property_desired_temperature_name
+    = AZ_SPAN_LITERAL_FROM_STR("targetTemperature");
+static az_span const property_reported_maximum_temperature_name
+    = AZ_SPAN_LITERAL_FROM_STR("maxTempSinceLastReboot");
+
+// Plug and Play Command Values
+static az_span const command_getMaxMinReport_name = AZ_SPAN_LITERAL_FROM_STR("getMaxMinReport");
+static az_span const command_max_temp_name = AZ_SPAN_LITERAL_FROM_STR("maxTemp");
+static az_span const command_min_temp_name = AZ_SPAN_LITERAL_FROM_STR("minTemp");
+static az_span const command_avg_temp_name = AZ_SPAN_LITERAL_FROM_STR("avgTemp");
+static az_span const command_start_time_name = AZ_SPAN_LITERAL_FROM_STR("startTime");
+static az_span const command_end_time_name = AZ_SPAN_LITERAL_FROM_STR("endTime");
+static az_span const command_empty_response_payload = AZ_SPAN_LITERAL_FROM_STR("{}");
+static char command_start_time_value_buffer[32];
+static char command_end_time_value_buffer[32];
+static char command_response_payload_buffer[256];
+
+// Plug and Play Telemetry Values
+static az_span const telemetry_temperature_name = AZ_SPAN_LITERAL_FROM_STR("temperature");
+
+// PnP Device Values
+static double device_current_temperature = DEFAULT_START_TEMP_CELSIUS;
+static double device_maximum_temperature = DEFAULT_START_TEMP_CELSIUS;
+static double device_minimum_temperature = DEFAULT_START_TEMP_CELSIUS;
+static double device_temperature_summation = DEFAULT_START_TEMP_CELSIUS;
+static uint32_t device_temperature_count = DEFAULT_START_TEMP_COUNT;
+static double device_average_temperature = DEFAULT_START_TEMP_CELSIUS;
+
+static iot_sample_environment_variables env_vars;
+static az_iot_pnp_client pnp_client;
+static az_iot_provisioning_client provisioning_client;
+static MQTTClient mqtt_client;
+static char iot_hub_endpoint_buffer[128];
+static char iot_hub_device_id_buffer[128];
+static az_span device_iot_hub_endpoint;
+static az_span device_id;
+static char mqtt_client_username_buffer[256];
+
+//
+// Provisioning Functions
+//
+static void create_and_configure_mqtt_client_for_provisioning(void);
+static void connect_mqtt_client_to_provisioning_service(void);
+static void subscribe_mqtt_client_to_provisioning_service_topics(void);
+static void register_device_with_provisioning_service(void);
+static void receive_device_registration_status_message(void);
+static void disconnect_mqtt_client_from_provisioning_service(void);
```

Comment:
There are a lot of functions in this sample. Worth adding doc comments for each (above the actual implementation) that explain what they are doing and why they are needed? As a reader, I am a bit lost, especially if I need to know which pieces to pick-and-choose.

Reply:
I think this can be better solved in the readme w/ added docs for provisioning. If provisioning is explained well enough in the docs, I think these function names make it pretty apparent their use.